src/main/java/com/commercetools/sync/products/utils/VariantReferenceReplacementUtils.java

```diff
@@ -0,0 +1,164 @@
+     * references. This method returns as a result a {@link List} of {@link AttributeDraft} that has all product
+     * references with keys replacing the ids.
+     *
+     * <p>Any product reference that is not expanded will have it's id in place and not replaced by the key.
+     *
+     * @param productVariant the product variant to replace its attribute product references ids with keys.
+     * @return  a {@link List} of {@link AttributeDraft} that has all product references with keys replacing the ids.
+     */
+    @Nonnull
+    static List<AttributeDraft> replaceAttributesReferencesIdsWithKeys(@Nonnull final ProductVariant productVariant) {
+        final List<Attribute> variantAttributes = productVariant.getAttributes();
+        final List<AttributeDraft> variantAttributeDraftsWithKeys = new ArrayList<>();
+        variantAttributes.forEach(attribute -> {
+            final AttributeDraft resolvedAttributeDraft = replaceAttributeReferenceIdWithKey(attribute)
+                .map(productReference -> AttributeDraft.of(attribute.getName(), productReference))
+                .orElseGet(() ->
+                    replaceAttributeReferenceSetIdsWithKeys(attribute)
+                        .map(productReferenceSet -> AttributeDraft.of(attribute.getName(), productReferenceSet))
+                        .orElseGet(() -> AttributeDraft.of(attribute.getName(), attribute.getValueAsJsonNode())));
+            variantAttributeDraftsWithKeys.add(resolvedAttributeDraft);
+        });
+        return variantAttributeDraftsWithKeys;
+    }
+
+
+    @SuppressWarnings("ConstantConditions") // NPE cannot occur due to being checked in replaceReferenceIdWithKey
+    static Optional<Reference<Product>> replaceAttributeReferenceIdWithKey(@Nonnull final Attribute attribute) {
+        return getProductReference(attribute)
+            .map(productReference -> replaceReferenceIdWithKey(productReference,
+                () -> Product.referenceOfId(productReference.getObj().getKey())));
+    }
+
+    private static Optional<Reference<Product>> getProductReference(@Nonnull final Attribute attribute) {
+        try {
+            return Optional.of(attribute.getValue(AttributeAccess.ofProductReference()));
+        } catch (final JsonException exception) {
+            return Optional.empty();
+        }
+    }
+
+    @SuppressWarnings("ConstantConditions") // NPE cannot occur due to being checked in replaceReferenceIdWithKey
+    static Optional<Set<Reference<Product>>> replaceAttributeReferenceSetIdsWithKeys(
+        @Nonnull final Attribute attribute) {
+        return getProductReferenceSet(attribute).map(productReferenceSet ->
+            productReferenceSet.stream()
+                               .map(productReference ->
+                                   replaceReferenceIdWithKey(productReference,
+                                       () -> Product.referenceOfId(productReference.getObj().getKey())))
```

Comment:
Usually it is better, instead of Supplier with closure, to have a normal function, e.g.:
`SyncUtils`:
```java

@Nullable
public static <T> Reference<T> replaceReferenceIdWithKey(
        @Nullable final Reference<T> reference,
        @Nonnull final Function<Reference<T>, Reference<T>> keyInReferenceSupplier) {
    if (reference != null && reference.getObj() != null) {
        return keyInReferenceSupplier.apply(reference);
    }
    return reference;
}
```
`VariantReferenceReplacementUtils`
```java
replaceReferenceIdWithKey(productReference,
        ref -> Product.referenceOfId(ref.getObj().getKey())))
```

Try to avoid closures as much as possible: they used to be slower and heavier (require more heap and stack to save environmen)

Reply:
@andrii-kovalenko-ct ok, I didn't know about this. I would appreciate if you provide a reference to read about this. Because I don't see any other reason to not use supplier here.

Reply:
well, i don't know any pure article about this, but this is just a common concept:
https://en.wikipedia.org/wiki/Closure_(computer_programming)

  - closures are created with _environment scope_, e.g. they _capture_ (copy) all the values they will be using when called, pretty much same like function calls put to stack everything they are depended on. That's why they are not absolutely _pure_: they capture and block values until they are removed. This might create heavy values in the memory, which are not called at all. In this certain case this closure captures _productReference_
  - opposite to above, pure functions don't capture environment, so they are more lightweight, more readable, more independent: when you read the function body you completely understand the logic and you don't need to understand in which environment the function is called - it depends only on input arguments.

In this certain case the difference (both from understandability and efficiency) is insignificant, this is only a recommendation for future.